common/types/encoding/codecv0/codecv0.go

```diff
@@ -0,0 +1,499 @@
+package codecv0
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	"strings"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/crypto"
+
+	"scroll-tech/common/types/encoding"
+)
+
+// CodecV0Version denotes the version of the codec.
+const CodecV0Version = 0
+
+// DABlock represents a Data Availability Block.
+type DABlock struct {
+	BlockNumber     uint64
+	Timestamp       uint64
+	BaseFee         *big.Int
+	GasLimit        uint64
+	NumTransactions uint16
+	NumL1Messages   uint16
+}
+
+// DAChunk groups consecutive DABlocks with their transactions.
+type DAChunk struct {
+	Blocks       []*DABlock
+	Transactions [][]*types.TransactionData
+}
+
+// DABatch contains metadata about a batch of DAChunks.
+type DABatch struct {
+	Version                uint8
+	BatchIndex             uint64
+	L1MessagePopped        uint64
+	TotalL1MessagePopped   uint64
+	DataHash               common.Hash
+	ParentBatchHash        common.Hash
+	SkippedL1MessageBitmap []byte
+}
+
+// NewDABlock creates a new DABlock from the given encoding.Block and the total number of L1 messages popped before.
+func NewDABlock(block *encoding.Block, totalL1MessagePoppedBefore uint64) (*DABlock, error) {
+	if !block.Header.Number.IsUint64() {
+		return nil, errors.New("block number is not uint64")
+	}
+
+	// note: numL1Messages includes skipped messages
+	numL1Messages := block.NumL1Messages(totalL1MessagePoppedBefore)
+	if numL1Messages > math.MaxUint16 {
```

Comment:
Is 65535 is too large?

Reply:
Because in the encoding, block context gives 2 bytes for the number of L1 messages, thus the number cannot be greater than 65536.

Reply:
https://github.com/scroll-tech/scroll/blob/v4.3.67/contracts/src/libraries/codec/ChunkCodec.sol#L8
https://github.com/scroll-tech/scroll/blob/v4.3.67/contracts/src/libraries/codec/ChunkCodec.sol#L24
https://github.com/scroll-tech/scroll/blob/v4.3.67/contracts/src/libraries/codec/ChunkCodec.sol#L25